Error: Another restore process of the same workspace is in progress

Got about 100 errors for 10 days in logs like:
```
2016-12-12 15:37:47,182[spaceManager-16]  [ERROR] [o.e.c.a.w.s.WorkspaceManager 660]    - org.eclipse.che.api.core.ServerException: Start of environment ****** failed. Error: Restore of workspace workspace111111111111 failed. Another restore process of the same workspace is in progress
java.util.concurrent.ExecutionException: org.eclipse.che.api.core.ServerException: Start of environment Curvytron failed. Error: Restore of workspace workspace1111111 failed. Another restore process of the same workspace is in progress
        at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[na:1.8.0_45]
        at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[na:1.8.0_45]
        at org.eclipse.che.api.workspace.server.WorkspaceManager.lambda$performAsyncStart$4(WorkspaceManager.java:639) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceManager$$Lambda$159/303085739.run(Unknown Source) [che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:28) [che-core-commons-lang-5.0.0-M8.jar:5.0.0-M8]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
Caused by: org.eclipse.che.api.core.ServerException: Start of environment Curvytron failed. Error: Restore of workspace workspace1111111 failed. Another restore process of the same workspace is in progress
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.doStart(WorkspaceRuntimes.java:665) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.lambda$startAsync$1(WorkspaceRuntimes.java:245) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$$Lambda$158/1742872442.call(Unknown Source) ~[na:na]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalCallable.call(CopyThreadLocalCallable.java:30) ~[che-core-commons-lang-5.0.0-M8.jar:5.0.0-M8]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_45]
        ... 3 common frames omitted
Caused by: org.eclipse.che.api.machine.server.exception.MachineException: Restore of workspace workspace11111111 failed. Another restore process of the same workspace is in progress
        at com.codenvy.machine.RemoteDockerNode.bindWorkspace(RemoteDockerNode.java:143) ~[codenvy-machine-hosted-5.0.0-M8.jar:5.0.0-M8]
        at com.codenvy.machine.agent.launcher.ExternalRsyncAgentLauncherImpl.launch(ExternalRsyncAgentLauncherImpl.java:63) ~[codenvy-machine-hosted-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.launchAgents(WorkspaceRuntimes.java:604) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.launchAgents(WorkspaceRuntimes.java:676) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.doStart(WorkspaceRuntimes.java:640) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        ... 7 common frames omitted
Caused by: org.eclipse.che.api.core.ServerException: Restore of workspace workspace111111111 failed. Another restore process of the same workspace is in progress
        at com.codenvy.machine.backup.MachineBackupManager.restoreWorkspaceBackup(MachineBackupManager.java:244) ~[codenvy-machine-hosted-5.0.0-M8.jar:5.0.0-M8]
        at com.codenvy.machine.RemoteDockerNode.bindWorkspace(RemoteDockerNode.java:133) ~[codenvy-machine-hosted-5.0.0-M8.jar:5.0.0-M8]
        ... 11 common frames omitted
```

**Reproduction Steps:**
Don't know, like run coednvy.io for several hours

**Expected behavior:**

No errors when restoring workspace.

**Observed behavior:**

Errors when restoring workspace.

**Codenvy version:** 5.0.0-M8
**OS and version:** Centos 7.2
**Additional information:**

Problem only started happening recently, didn't happen in an older version of Codenvy: Yes
Problem can be reliably reproduced: No - it's random

Receiving errors with slightly different in content in 5.2.2
```
2017-02-06 17:19:55,111[ceSharedPool-11]  [ERROR] [DockerEnvironmentBackupManager 313]  - Restore of workspace workspace11111111111111 failed. Another restore process of the same workspace is in progress
2017-02-06 17:19:55,112[ceSharedPool-11]  [WARN ] [c.c.machine.RemoteDockerNode 101]    - The container '222222222222222222' in the workspace 'workspace111111111111' won't be backed up. The reason is that the workspace wasn't restored from backup
...
2017-02-06 17:19:55,905[ceSharedPool-11]  [ERROR] [o.e.c.a.w.s.WorkspaceManager 774]    - Restore of workspace workspace1111111111111 failed. Another restore process of the same workspace is in progress
org.eclipse.che.api.core.ServerException: Restore of workspace workspace111111111111 failed. Another restore process of the same workspace is in progress
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.restoreBackupInsideLock(DockerEnvironmentBackupManager.java:314) ~[codenvy-machine-hosted-5.2.2.jar:5.2.2]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.restoreWorkspaceBackup(DockerEnvironmentBackupManager.java:216) ~[codenvy-machine-hosted-5.2.2.jar:5.2.2]
        at com.codenvy.machine.RemoteDockerNode.bindWorkspace(RemoteDockerNode.java:94) ~[codenvy-machine-hosted-5.2.2.jar:5.2.2]
        at com.codenvy.machine.agent.launcher.ExternalRsyncAgentLauncherImpl.launch(ExternalRsyncAgentLauncherImpl.java:68) ~[codenvy-machine-hosted-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.launchAgents(WorkspaceRuntimes.java:693) ~[che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$MachineAgentsLauncher.started(WorkspaceRuntimes.java:1054) ~[che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:806) ~[che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.start(CheEnvironmentEngine.java:260) ~[che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.startEnvironmentAndPublishEvents(WorkspaceRuntimes.java:712) ~[che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.access$100(WorkspaceRuntimes.java:105) ~[che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartTask.call(WorkspaceRuntimes.java:977) [che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartTask.call(WorkspaceRuntimes.java:941) [che-core-api-workspace-5.2.2.jar:5.2.2]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalCallable.call(CopyThreadLocalCallable.java:30) ~[che-core-commons-lang-5.2.2.jar:5.2.2]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_102]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_102]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_102]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_102]
```

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/DockerEnvironmentBackupManager.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/DockerEnvironmentBackupManager.java
@@ -35,7 +35,6 @@ import org.eclipse.che.plugin.docker.client.json.PortBinding;
 import org.eclipse.che.plugin.docker.client.params.CreateExecParams;
 import org.eclipse.che.plugin.docker.client.params.StartExecParams;
 import org.eclipse.che.plugin.docker.machine.DockerInstance;
-import org.eclipse.che.plugin.docker.machine.node.WorkspaceFolderPathProvider;
 import org.slf4j.Logger;
 
 import javax.inject.Inject;
@@ -53,7 +52,6 @@ import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.locks.ReentrantLock;
 
-import static com.codenvy.machine.agent.CodenvyInfrastructureProvisioner.INFRASTRUCTURE_TYPE_PROPERTY;
 import static com.google.common.base.MoreObjects.firstNonNull;
 import static java.lang.String.format;
 import static java.util.Collections.emptyMap;
@@ -80,9 +78,7 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
     private final WorkspaceIdHashLocationFinder        workspaceIdHashLocationFinder;
     private final String                               projectFolderPath;
     private final ConcurrentMap<String, ReentrantLock> workspacesBackupLocks;
-    private final boolean                              syncAgentInMachine;
     private final WorkspaceManager                     workspaceManager;
-    private final WorkspaceFolderPathProvider          workspaceFolderPathProvider;
     private final DockerConnector                      dockerConnector;
 
     @Inject
@@ -92,10 +88,8 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
                                           @Named("machine.backup.restore_duration_second") int restoreDurationSec,
                                           @Named("che.user.workspaces.storage") File backupsRootDir,
                                           WorkspaceIdHashLocationFinder workspaceIdHashLocationFinder,
-                                          @Named(INFRASTRUCTURE_TYPE_PROPERTY) String syncStrategy,
                                           @Named("che.workspace.projects.storage") String projectFolderPath,
                                           WorkspaceManager workspaceManager,
-                                          WorkspaceFolderPathProvider workspaceFolderPathProvider,
                                           DockerConnector dockerConnector) {
         this.backupScript = backupScript;
         this.restoreScript = restoreScript;
@@ -105,22 +99,7 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
         this.workspaceIdHashLocationFinder = workspaceIdHashLocationFinder;
         this.projectFolderPath = projectFolderPath;
         this.workspaceManager = workspaceManager;
-        this.workspaceFolderPathProvider = workspaceFolderPathProvider;
         this.dockerConnector = dockerConnector;
-
-        switch (syncStrategy) {
-            case "native":
-                syncAgentInMachine = false;
-                break;
-            case "in-container":
-                syncAgentInMachine = true;
-                break;
-            default:
-                throw new RuntimeException(format(
-                        "Property '%s' has illegal value '%s'. Valid values: native, in-container",
-                        INFRASTRUCTURE_TYPE_PROPERTY,
-                        syncStrategy));
-        }
 
         workspacesBackupLocks = new ConcurrentHashMap<>();
     }
@@ -141,13 +120,12 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
                                                                                                   devMachine.getId());
             // machine that is not in running state can be just a stub and should not be casted
             String nodeHost = dockerDevMachine.getNode().getHost();
-            String srcPath = syncAgentInMachine ? projectFolderPath : workspaceFolderPathProvider.getPath(workspaceId);
             String destPath = workspaceIdHashLocationFinder.calculateDirPath(backupsRootDir, workspaceId).toString();
             String srcUserName = getUserName(workspaceId, dockerDevMachine.getContainer()).name;
             int syncPort = getSyncPort(dockerDevMachine);
 
             backupInsideLock(workspaceId,
-                             srcPath,
+                             projectFolderPath,
                              nodeHost,
                              syncPort,
                              srcUserName,
@@ -173,21 +151,23 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
                                           String containerId,
                                           String nodeHost) throws ServerException {
         try {
-            // TODO move to another class to avoid conditional logic here
-            String srcPath = syncAgentInMachine ? projectFolderPath : workspaceFolderPathProvider.getPath(workspaceId);
             String destPath = workspaceIdHashLocationFinder.calculateDirPath(backupsRootDir, workspaceId).toString();
             // if sync agent is not in machine port parameter is not used
             int syncPort = getSyncPort(containerId);
             String srcUserName = getUserName(workspaceId, containerId).name;
 
             backupAndCleanupInsideLock(workspaceId,
-                                       srcPath,
+                                       projectFolderPath,
                                        nodeHost,
                                        syncPort,
                                        srcUserName,
                                        destPath);
         } catch (IOException e) {
             throw new ServerException(e.getLocalizedMessage(), e);
+        } finally {
+            // remove lock in case exception prevent removing it in regular place to prevent resources leak
+            // and blocking further WS start
+            removeWorkspaceLock(workspaceId);
         }
     }
 
@@ -208,14 +188,12 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
                                        String nodeHost) throws ServerException {
         try {
             String srcPath = workspaceIdHashLocationFinder.calculateDirPath(backupsRootDir, workspaceId).toString();
-            // TODO move to another class to avoid conditional logic here
-            String destPath = syncAgentInMachine ? projectFolderPath : workspaceFolderPathProvider.getPath(workspaceId);
             User user = getUserNameAndIds(workspaceId, containerId);
             int syncPort = getSyncPort(containerId);
 
             restoreBackupInsideLock(workspaceId,
                                     srcPath,
-                                    destPath,
+                                    projectFolderPath,
                                     user.id,
                                     user.groupId,
                                     user.name,
@@ -287,8 +265,7 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
                                     srcUserName,
                                     destPath);
             } finally {
-                workspacesBackupLocks.remove(workspaceId);
-                lock.unlock();
+                removeWorkspaceLock(workspaceId);
             }
         } else {
             LOG.warn("Attempt to backup workspace {} after cleanup", workspaceId);
@@ -379,6 +356,13 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
             throw new ServerException(
                     "Backup of workspace " + workspaceId + " filesystem terminated on " + srcAddress + " node. "
                     + e.getLocalizedMessage());
+        }
+    }
+
+    private void removeWorkspaceLock(String workspaceId) {
+        ReentrantLock lock = workspacesBackupLocks.remove(workspaceId);
+        if (lock != null && lock.isLocked()) {
+            lock.unlock();
         }
     }
 
@@ -482,13 +466,7 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
      * @throws ServerException
      *         if port is not found
      */
-    // TODO move to another class to avoid conditional logic here
     private int getSyncPort(String containerId) throws ServerException {
-        if (!syncAgentInMachine) {
-            // in that case port doesn't matter
-            return 0;
-        }
-
         ContainerInfo containerInfo;
         try {
             containerInfo = dockerConnector.inspectContainer(containerId);
@@ -514,14 +492,7 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
      * @throws ServerException
      *         if port is not found
      */
-    // TODO move to another class to avoid conditional logic here
-    // TODO maybe use only version with containerID?
     private int getSyncPort(Machine machine) throws ServerException {
-        if (!syncAgentInMachine) {
-            // in that case port doesn't matter
-            return 0;
-        }
-
         Server server = machine.getRuntime().getServers().get("22/tcp");
         if (server == null) {
             throw new ServerException(

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/DockerEnvironmentBackupManager.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/DockerEnvironmentBackupManager.java
@@ -167,7 +167,7 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
         } finally {
             // remove lock in case exception prevent removing it in regular place to prevent resources leak
             // and blocking further WS start
-            removeWorkspaceLock(workspaceId);
+            workspacesBackupLocks.remove(workspaceId);
         }
     }
 
@@ -265,7 +265,8 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
                                     srcUserName,
                                     destPath);
             } finally {
-                removeWorkspaceLock(workspaceId);
+                workspacesBackupLocks.remove(workspaceId);
+                lock.unlock();
             }
         } else {
             LOG.warn("Attempt to backup workspace {} after cleanup", workspaceId);
@@ -356,13 +357,6 @@ public class DockerEnvironmentBackupManager implements EnvironmentBackupManager 
             throw new ServerException(
                     "Backup of workspace " + workspaceId + " filesystem terminated on " + srcAddress + " node. "
                     + e.getLocalizedMessage());
-        }
-    }
-
-    private void removeWorkspaceLock(String workspaceId) {
-        ReentrantLock lock = workspacesBackupLocks.remove(workspaceId);
-        if (lock != null && lock.isLocked()) {
-            lock.unlock();
         }
     }
 

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/backup/DockerEnvironmentBackupManagerTest.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/backup/DockerEnvironmentBackupManagerTest.java
@@ -19,6 +19,7 @@ import org.eclipse.che.api.core.ServerException;
 import org.eclipse.che.api.core.model.machine.MachineStatus;
 import org.eclipse.che.api.machine.server.model.impl.MachineImpl;
 import org.eclipse.che.api.machine.server.model.impl.MachineRuntimeInfoImpl;
+import org.eclipse.che.api.machine.server.model.impl.ServerImpl;
 import org.eclipse.che.api.workspace.server.WorkspaceManager;
 import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
 import org.eclipse.che.api.workspace.server.model.impl.WorkspaceRuntimeImpl;
@@ -26,11 +27,13 @@ import org.eclipse.che.plugin.docker.client.DockerConnector;
 import org.eclipse.che.plugin.docker.client.Exec;
 import org.eclipse.che.plugin.docker.client.LogMessage;
 import org.eclipse.che.plugin.docker.client.MessageProcessor;
+import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
+import org.eclipse.che.plugin.docker.client.json.NetworkSettings;
+import org.eclipse.che.plugin.docker.client.json.PortBinding;
 import org.eclipse.che.plugin.docker.client.params.CreateExecParams;
 import org.eclipse.che.plugin.docker.client.params.StartExecParams;
 import org.eclipse.che.plugin.docker.machine.DockerInstance;
 import org.eclipse.che.plugin.docker.machine.node.DockerNode;
-import org.eclipse.che.plugin.docker.machine.node.WorkspaceFolderPathProvider;
 import org.mockito.ArgumentCaptor;
 import org.mockito.Captor;
 import org.mockito.Matchers;
@@ -53,6 +56,8 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.ReentrantLock;
 
 import static java.lang.Thread.sleep;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
 import static org.mockito.Matchers.any;
 import static org.mockito.Matchers.anyInt;
 import static org.mockito.Matchers.anyObject;
@@ -87,43 +92,40 @@ public class DockerEnvironmentBackupManagerTest {
     private static final int    MAX_BACKUP_DURATION_SEC        = 10;
     private static final int    MAX_RESTORE_DURATION_SEC       = 10;
     private static final String BACKUPS_ROOT_PATH              = "/tmp/che/backups";
-    private static final String SRC_PATH                       = "/some/path/on/docker/node";
     private static final String USER_ID                        = "1234";
     private static final String USER_GID                       = "12345";
     private static final String PATH_TO_WORKSPACE              = "00/00/00/";
     private static final String ABSOLUTE_PATH_TO_WORKSPACE_DIR = BACKUPS_ROOT_PATH + PATH_TO_WORKSPACE + WORKSPACE_ID;
     private static final String NODE_HOST                      = "192.19.20.78";
-    private static final String SYNC_STRATEGY                  = "native";
     private static final String PROJECTS_PATH_IN_CONTAINER     = "/projects-folder";
     private static final String USER_IN_CONTAINER              = "test-user";
+    private static final String PUBLISHED_SSH_PORT             = "32559";
 
     private static final String[] BACKUP_WORKSPACE_COMMAND              = {BACKUP_SCRIPT,
-                                                                           SRC_PATH,
+                                                                           PROJECTS_PATH_IN_CONTAINER,
                                                                            NODE_HOST,
-                                                                           "0",
+                                                                           PUBLISHED_SSH_PORT,
                                                                            ABSOLUTE_PATH_TO_WORKSPACE_DIR,
                                                                            "false",
                                                                            USER_IN_CONTAINER};
     private static final String[] BACKUP_WORKSPACE_WITH_CLEANUP_COMMAND = {BACKUP_SCRIPT,
-                                                                           SRC_PATH,
+                                                                           PROJECTS_PATH_IN_CONTAINER,
                                                                            NODE_HOST,
-                                                                           "0",
+                                                                           PUBLISHED_SSH_PORT,
                                                                            ABSOLUTE_PATH_TO_WORKSPACE_DIR,
                                                                            "true",
                                                                            USER_IN_CONTAINER};
     private static final String[] RESTORE_WORKSPACE_COMMAND             = {RESTORE_SCRIPT,
                                                                            ABSOLUTE_PATH_TO_WORKSPACE_DIR,
-                                                                           SRC_PATH,
+                                                                           PROJECTS_PATH_IN_CONTAINER,
                                                                            NODE_HOST,
-                                                                           "0",
+                                                                           PUBLISHED_SSH_PORT,
                                                                            USER_ID,
                                                                            USER_GID,
                                                                            USER_IN_CONTAINER};
 
     @Mock
     private WorkspaceIdHashLocationFinder workspaceIdHashLocationFinder;
-    @Mock
-    private WorkspaceFolderPathProvider   workspaceFolderPathProvider;
     @Mock
     private DockerConnector               docker;
     @Mock
@@ -156,10 +158,8 @@ public class DockerEnvironmentBackupManagerTest {
                                                                MAX_RESTORE_DURATION_SEC,
                                                                new File(BACKUPS_ROOT_PATH),
                                                                workspaceIdHashLocationFinder,
-                                                               SYNC_STRATEGY,
                                                                PROJECTS_PATH_IN_CONTAINER,
                                                                workspaceManager,
-                                                               workspaceFolderPathProvider,
                                                                docker));
 
         when(workspaceManager.getWorkspace(anyString())).thenReturn(workspace);
@@ -171,7 +171,9 @@ public class DockerEnvironmentBackupManagerTest {
         when(dockerNode.getHost()).thenReturn(NODE_HOST);
         when(dockerInstance.getContainer()).thenReturn(CONTAINER_ID);
         when(dockerInstance.getRuntime()).thenReturn(machineRuntimeInfo);
-        when(workspaceFolderPathProvider.getPath(WORKSPACE_ID)).thenReturn(SRC_PATH);
+        when(machineRuntimeInfo.getServers())
+                .thenReturn(singletonMap("22/tcp", new ServerImpl("ref", "proto", "127.0.0.1:" + PUBLISHED_SSH_PORT,
+                                                                  null, null)));
         when(workspaceIdHashLocationFinder.calculateDirPath(any(File.class), any(String.class)))
                 .thenReturn(new File(ABSOLUTE_PATH_TO_WORKSPACE_DIR));
         doNothing().when(backupManager).executeCommand(anyObject(), anyInt(), anyString(), anyString());
@@ -203,6 +205,12 @@ public class DockerEnvironmentBackupManagerTest {
 
             return null;
         }).when(docker).startExec(any(StartExecParams.class), Matchers.any());
+        NetworkSettings networkSettings = new NetworkSettings();
+        networkSettings.setPorts(singletonMap("22/tcp",
+                                              singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
+        ContainerInfo containerInfo = new ContainerInfo();
+        containerInfo.setNetworkSettings(networkSettings);
+        when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
 
         executor = Executors.newFixedThreadPool(5);
     }
@@ -250,7 +258,6 @@ public class DockerEnvironmentBackupManagerTest {
                                  dockerNode,
                                  machineRuntimeInfo,
                                  devMachine,
-                                 workspaceFolderPathProvider,
                                  workspaceIdHashLocationFinder);
     }
 
@@ -268,7 +275,6 @@ public class DockerEnvironmentBackupManagerTest {
                                  dockerNode,
                                  workspaceManager,
                                  machineRuntimeInfo,
-                                 workspaceFolderPathProvider,
                                  workspaceIdHashLocationFinder);
     }
 
@@ -288,6 +294,22 @@ public class DockerEnvironmentBackupManagerTest {
         backupManager.backupWorkspace(WORKSPACE_ID);
 
         verify(backupManager, never()).executeCommand(any(String[].class), anyInt(), anyString(), anyString());
+    }
+
+    @Test(expectedExceptions = ServerException.class,
+          expectedExceptionsMessageRegExp = "Sync port is not exposed in ws-machine. Workspace projects syncing is not possible")
+    public void backupShouldThrowErrorIfMachineDoesNotContainServerWithSshPort() throws Exception {
+        // given
+        when(machineRuntimeInfo.getServers())
+                .thenReturn(singletonMap("23/tcp", new ServerImpl("ref", "proto", "127.0.0.1:" + PUBLISHED_SSH_PORT,
+                                                                  null, null)));
+
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+
+        // when
+        backupManager.backupWorkspace(WORKSPACE_ID);
     }
 
     @Test
@@ -476,6 +498,23 @@ public class DockerEnvironmentBackupManagerTest {
         assertEquals(cmdCaptor.getValue()[0], BACKUP_SCRIPT);
     }
 
+    @Test(expectedExceptions = ServerException.class,
+          expectedExceptionsMessageRegExp = "Sync port is not exposed in ws-machine. Workspace projects syncing is not possible")
+    public void restoreShouldThrowErrorIfSshPortIsNotPublishedInContainer() throws Exception {
+        // given
+        NetworkSettings networkSettings = new NetworkSettings();
+        networkSettings.setPorts(singletonMap("23/tcp",
+                                              singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
+        ContainerInfo containerInfo = new ContainerInfo();
+        containerInfo.setNetworkSettings(networkSettings);
+        when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
+
+        // when
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+    }
+
     @Test
     public void shouldIgnoreOthersBackupsAndFailOnRestoresWhileBackupInProgress() throws Exception {
         // given
@@ -593,6 +632,48 @@ public class DockerEnvironmentBackupManagerTest {
                                                 CONTAINER_ID,
                                                 NODE_HOST);
         verify(backupManager, times(2)).executeCommand(anyObject(), anyInt(), eq(NODE_HOST), anyString());
+    }
+
+    @Test
+    public void shouldBeAbleToRestoreAfterBackupWithCleanupWhenExceptionIsThrown() throws Exception {
+        // given
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+        when(workspaceIdHashLocationFinder.calculateDirPath(any(File.class), anyString()))
+                .thenThrow(new RuntimeException("some test error"))
+                .thenReturn(new File(ABSOLUTE_PATH_TO_WORKSPACE_DIR));
+        try {
+            backupManager.backupWorkspaceAndCleanup(WORKSPACE_ID,
+                                                    CONTAINER_ID,
+                                                    NODE_HOST);
+        } catch (RuntimeException ignored) {}
+
+        // when
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+    }
+
+    @Test(expectedExceptions = ServerException.class,
+          expectedExceptionsMessageRegExp = "Sync port is not exposed in ws-machine. Workspace projects syncing is not possible")
+    public void cleanupShouldThrowErrorIfSshPortIsNotPublishedInContainer() throws Exception {
+        // given
+        NetworkSettings networkSettings = new NetworkSettings();
+        networkSettings.setPorts(singletonMap("23/tcp",
+                                              singletonList(new PortBinding().withHostPort(PUBLISHED_SSH_PORT))));
+        ContainerInfo containerInfo = new ContainerInfo();
+        containerInfo.setNetworkSettings(networkSettings);
+        when(docker.inspectContainer(eq(CONTAINER_ID))).thenReturn(containerInfo);
+
+        backupManager.restoreWorkspaceBackup(WORKSPACE_ID,
+                                             CONTAINER_ID,
+                                             NODE_HOST);
+
+        // when
+        backupManager.backupWorkspaceAndCleanup(WORKSPACE_ID,
+                                                CONTAINER_ID,
+                                                NODE_HOST);
     }
 
     @Test

